test/Feature/SingularSpec.hs

```diff
@@ -49,48 +48,52 @@ spec =
 
     context "when updating rows" $ do
       it "works for one row with return=rep" $ do
-        post "/addresses" [json| { id: 97, address: "A Street" } |]
-        request methodPatch "/addresses?id=eq.97"
-           [("Prefer", "return=representation"), singular]
-           [json| { address: "B Street" } |]
+        request methodPatch "/addresses?id=eq.1"
+            [("Prefer", "return=representation"), singular]
+            [json| { address: "B Street" } |]
           `shouldRespondWith`
-            [json|{"id":97,"address":"B Street"}|]
+            [json|{"id":1,"address":"B Street"}|]
             { matchHeaders = [matchContentTypeSingular] }
 
       it "works for one row with return=minimal" $
-        request methodPatch
-          "/addresses?id=eq.97"
-          [("Prefer", "return=minimal"), singular]
-          [json| { address: "C Street" } |]
+        request methodPatch "/addresses?id=eq.1"
+            [("Prefer", "return=minimal"), singular]
+            [json| { address: "C Street" } |]
           `shouldRespondWith`
             ""
             { matchStatus  = 204 }
 
       it "raises an error for multiple rows" $ do
-        _ <- post "/addresses" [json| { id: 98, address: "xxx" } |]
-        _ <- post "/addresses" [json| { id: 99, address: "yyy" } |]
-        p <- request methodPatch "/addresses?id=gt.0"
-                [singular]
-                [json| { address: "zzz" } |]
-        liftIO $ do
-          simpleStatus p `shouldBe` notAcceptable406
-          isErrorFormat (simpleBody p) `shouldBe` True
+        request methodPatch "/addresses"
+            [("Prefer", "tx=commit"), singular]
+            [json| { address: "zzz" } |]
+          `shouldRespondWith`
+            [json|{"details":"Results contain 4 rows, application/vnd.pgrst.object+json requires 1 row","message":"JSON object requested, multiple (or no) rows returned"}|]
+            { matchStatus  = 406
+            , matchHeaders = [ matchContentTypeSingular
+                             , "Preference-Applied" <:> "tx=commit" ]
+            }
 
         -- the rows should not be updated, either
-        get "/addresses?id=eq.98" `shouldRespondWith` [json|[{"id":98,"address":"xxx"}]|]
+        get "/addresses?id=eq.1"
+          `shouldRespondWith`
+            [json|[{"id":1,"address":"address 1"}]|]
 
       it "raises an error for multiple rows with return=rep" $ do
-        _ <- post "/addresses" [json| { id: 100, address: "xxx" } |]
-        _ <- post "/addresses" [json| { id: 101, address: "yyy" } |]
-        p <- request methodPatch "/addresses?id=gt.0"
-                [("Prefer", "return=representation"), singular]
-                [json| { address: "zzz" } |]
-        liftIO $ do
-          simpleStatus p `shouldBe` notAcceptable406
-          isErrorFormat (simpleBody p) `shouldBe` True
+        request methodPatch "/addresses"
+            [("Prefer", "tx=commit"), ("Prefer", "return=representation"), singular]
```

Comment:
Some tests that were expected to fail, but then had additional `get` requests to confirm that no data was changed, I had to add a `Prefer tx=commit` header. I also check explicitely for `Preference-Applied` to make sure that the "non-persistance" is because of the errored request and not because of a rollback.

Reply:
Ah.. so for testing errors we should add the `Prefer: tx=commit`. I'll take note.

Reply:
Not all errors need `Prefer: tx=commit`. Once you test for "non-persistance after an error" you need it. Most tests for errors do not have the "non-persistance test" part.